hibernate-core/src/main/java/org/hibernate/bytecode/enhance/internal/bytebuddy/InlineDirtyCheckingHandler.java

```diff
@@ -96,6 +101,93 @@ public Size apply(
+					methodVisitor.visitInsn( Opcodes.LCONST_0 );
+					methodVisitor.visitInsn( Opcodes.LCMP );
+					methodVisitor.visitJumpInsn( Opcodes.IFNE, skipUninitializedCheck );
+				}
+				else if ( persistentField.getType().represents( float.class ) ) {
+					methodVisitor.visitInsn( Opcodes.FCONST_0 );
+					methodVisitor.visitInsn( Opcodes.FCMPL );
+					methodVisitor.visitJumpInsn( Opcodes.IFNE, skipUninitializedCheck );
+				}
+				else if ( persistentField.getType().represents( double.class ) ) {
+					methodVisitor.visitInsn( Opcodes.DCONST_0 );
+					methodVisitor.visitInsn( Opcodes.DCMPL );
+					methodVisitor.visitJumpInsn( Opcodes.IFNE, skipUninitializedCheck );
+				}
+				else {
+					methodVisitor.visitJumpInsn( Opcodes.IFNE, skipUninitializedCheck );
+				}
+			}
+			else {
+				methodVisitor.visitJumpInsn( Opcodes.IFNONNULL, skipUninitializedCheck );
+			}
+
+			// if ( this.$$_hibernate_getInterceptor() != null )
+			methodVisitor.visitVarInsn( Opcodes.ALOAD, 0 );
+			methodVisitor.visitMethodInsn(
+					Opcodes.INVOKEVIRTUAL,
+					managedCtClass.getInternalName(),
+					EnhancerConstants.INTERCEPTOR_GETTER_NAME,
+					Type.getMethodDescriptor( Type.getType( PersistentAttributeInterceptor.class ) ),
+					false
+			);
+
+			methodVisitor.visitInsn( Opcodes.DUP );
+
+			// if ( this.$$_hibernate_getInterceptor() == null ) => no dirty check
+			methodVisitor.visitJumpInsn( Opcodes.IFNULL, skipUninitializedCheckWithPersistentAttributeInterceptor );
+
+			// if ( !this.$$_hibernate_getInterceptor() instanceof LazyAttributeLoadingInterceptor ) => no dirty check
+			methodVisitor.visitInsn( Opcodes.DUP );
+			methodVisitor.visitTypeInsn( Opcodes.INSTANCEOF, Type.getInternalName( LazyAttributeLoadingInterceptor.class ) );
+			methodVisitor.visitJumpInsn( Opcodes.IFEQ, skipUninitializedCheckWithPersistentAttributeInterceptor );
+			methodVisitor.visitTypeInsn( Opcodes.CHECKCAST, Type.getInternalName( LazyAttributeLoadingInterceptor.class ) );
+
+			// call interceptor.isAttributeLoaded( fieldName )
+			methodVisitor.visitLdcInsn( persistentField.getName() );
+			methodVisitor.visitMethodInsn(
+					Opcodes.INVOKEVIRTUAL,
+					Type.getInternalName( LazyAttributeLoadingInterceptor.class ),
+					"isAttributeLoaded",
+					Type.getMethodDescriptor(
+							Type.BOOLEAN_TYPE,
+							Type.getType( String.class )
+					),
+					false
+			);
+
+			// When field is not initialized, comparison is useless
+			// if ( interceptor.isAttributeLoaded( fieldName ) ) goto skipUninitializedCheck
+			methodVisitor.visitJumpInsn( Opcodes.IFNE, skipUninitializedCheck );
+
+			// Sure that the lazy is not initialized. No need for comparison, directly set dirty
+			methodVisitor.visitJumpInsn( Opcodes.GOTO, uninitializedLazy );
+
+			methodVisitor.visitLabel( skipUninitializedCheckWithPersistentAttributeInterceptor );
+			if ( implementationContext.getClassFileVersion().isAtLeast( ClassFileVersion.JAVA_V6 ) ) {
```

Comment:
Could you comment about the class version requirement?

I'm also wondering if we should test this on pre-6 java class versions, but that's going to be complicated with our CI setup as the main codebase requires Java 8.

Reply:
Stackmap bytecodes have been introduced in JAVA6 : https://stackoverflow.com/questions/25109942/what-is-a-stack-map-frame.
I just followed already existing pattern here (there was already that kind of test near the end).